sys/kern/src/descs.rs

```diff
@@ -0,0 +1,149 @@
+    ///
+    /// Note: because these region indices are 8 bits, this is going to get
+    /// restrictive in applications that approach 128 tasks.
+    pub regions: [&'static RegionDesc; REGIONS_PER_TASK],
+    /// Address of the task's entry point. This is the first instruction that
+    /// will be executed whenever the task is (re)started. It must be within one
+    /// of the task's memory regions (the kernel *will* check this).
+    pub entry_point: u32,
+    /// Address of the task's initial stack pointer, to be loaded at (re)start.
+    /// It must be pointing into or *just past* one of the task's memory
+    /// regions (the kernel *will* check this).
+    pub initial_stack: u32,
+    /// Initial priority of this task.
+    pub priority: u8,
+    /// Collection of boolean flags controlling task behavior.
+    pub flags: TaskFlags,
+    /// Index of this task within the task table.
+    ///
+    /// This field is here as an optimization for the kernel entry sequences. It
+    /// can contain an invalid index if you create an arbitrary invalid
+    /// `TaskDesc`; this will cause the kernel to behave strangely (if the index
+    /// is in range for the task table) or panic predictably (if not), but won't
+    /// violate safety. The build system is careful to generate correct indices
+    /// here.
+    ///
+    /// The index is a u16 to save space in the `TaskDesc` struct; in practice
+    /// other factors limit us to fewer than `2**16` tasks.
+    pub index: u16,
+}
+
+bitflags::bitflags! {
+    #[repr(transparent)]
+    pub struct TaskFlags: u8 {
+        const START_AT_BOOT = 1 << 0;
+        const RESERVED = !1;
+    }
+}
+
+/// Description of one memory region.
+///
+/// A memory region can be used by multiple tasks. This is mostly used to have
+/// tasks share a no-access region (often index 0) in unused region slots, but
+/// you could also use it for shared peripheral or RAM access.
+///
+/// Note that regions can overlap. This can be useful: for example, you can have
+/// two regions pointing to the same area of the address space, but one
+/// read-only and the other read-write.
+#[derive(Clone, Debug)]
+pub struct RegionDesc {
+    /// Address of start of region. The platform likely has alignment
+    /// requirements for this; it must meet them. (For example, on ARMv7-M, it
+    /// must be naturally aligned for the size.)
+    pub base: u32,
+    /// Size of region, in bytes. The platform likely has alignment requirements
+    /// for this; it must meet them. (For example, on ARMv7-M, it must be a
+    /// power of two greater than 16.)
+    pub size: u32,
+    /// Flags describing what can be done with this region.
+    pub attributes: RegionAttributes,
+}
+
+impl RegionDesc {
+    /// Tests whether `self` contains `addr`.
+    pub fn contains(&self, addr: usize) -> bool {
+        let next_addr = addr.wrapping_add(1);
+        if next_addr < addr {
```

Comment:
Is this equivalent to `addr == usize::MAX`?

Reply:
It is _in this commit_ but I have an upcoming commit switching this to `*const T`, where the meaning of the addition will now depend on sizeof T. So that's why it's phrased this way.